packages/gatsby/src/schema/resolvers.ts

```diff
@@ -0,0 +1,509 @@
+        context.stats.totalPluralRunQuery++
+      }
+    }
+
+    const result = await context.nodeModel.runQuery(
+      {
+        query: runQueryArgs,
+        firstOnly,
+        type,
+        stats: context.stats,
+        tracer: context.tracer,
+      },
+      { path: context.path }
+    )
+    if (
+      returnType instanceof GraphQLList &&
+      Array.isArray(fieldValue) &&
+      Array.isArray(result)
+    ) {
+      return fieldValue.map(value =>
+        result.find(obj => getValueAt(obj, options.by) === value)
+      )
+    } else {
+      return result
+    }
+  }
+}
+
+export function fileByPath<TSource, TArgs>(
+  options: {
+    from?: string
+    fromNode?: string
+  } = {},
+  fieldConfig
+): GatsbyResolver<TSource, TArgs> {
+  return async function fileByPathResolver(
+    source,
+    args,
+    context,
+    info
+  ): Promise<any> {
+    const resolver = fieldConfig.resolve || context.defaultFieldResolver
+    const fieldValue = await resolver(source, args, context, {
+      ...info,
+      from: options.from || info.from,
+      fromNode: options.from ? options.fromNode : info.fromNode,
+    })
+
+    if (fieldValue == null) return null
+
+    // Find the File node for this node (we assume the node is something
+    // like markdown which would be a child node of a File node).
+    const parentFileNode = context.nodeModel.findRootNodeAncestor(
+      source,
+      node => node.internal && node.internal.type === `File`
+    )
+
+    const findLinkedFileNode = (relativePath: string): any => {
+      // Use the parent File node to create the absolute path to
+      // the linked file.
+      const fileLinkPath = normalize(
+        systemPath.resolve(parentFileNode.dir, relativePath)
+      )
+
+      // Use that path to find the linked File node.
+      const linkedFileNode = _.find(
+        context.nodeModel.getAllNodes({ type: `File` }),
+        n => n.absolutePath === fileLinkPath
+      )
+      return linkedFileNode
+    }
+
+    return resolveValue(findLinkedFileNode, fieldValue)
+  }
+}
+
+function resolveValue(
+  resolve: (a: any) => any,
+  value: any | Array<any>
+): any | Array<any> {
+  return Array.isArray(value)
+    ? value.map(v => resolveValue(resolve, v))
+    : resolve(value)
+}
+
+function getProjectedField(
+  info: GraphQLResolveInfo,
+  fieldName: string
+): Array<string> {
+  const selectionSet = info.fieldNodes[0].selectionSet
+  if (selectionSet) {
+    const fieldNodes = getFieldNodeByNameInSelectionSet(
+      selectionSet,
+      fieldName,
+      info
+    )
+
+    const returnType = getNullableType(info.returnType)
+
+    if (isObjectType(returnType) || isInterfaceType(returnType)) {
+      const field = returnType.getFields()[fieldName]
+      if (field) {
+        const fieldArg = field.args.find(arg => arg.name === `field`)
+        if (fieldArg) {
+          const fieldEnum = getNullableType(fieldArg.type)
+
+          if (fieldEnum && isEnumType(fieldEnum)) {
```

Comment:
Typescript hell above 🤯 Not sure if we can make it more readable - maybe empty defaults can help. Previous (pure js) version was easier to follow %)

Reply:
Made a bit cleaner.